### PR TITLE
[RFC/TESTS] Simplify/explain hda ipc

### PR DESCRIPTION
**Pull request by plbossart**
These two patches show that some of the current code is unnecessary.
1. the tests for HIPCCTL bits make no sense.
2. the second patch illustrates that interrupt masking is not necessary as long as the interrupt is not re-enabled before the BUSY/DONE bits are cleared. It seems the interrupt handling is level based and a new interrupt will be generated if those bits are not cleared first. I don't mind keeping the existing mechanism since it can speed up the response time to an unsolicited firmware message: we share the same interrupt routine for two potential sources, but I'd like to get comments on whether this can actually happen. If the firmware is totally serial, and cannot deal with interleaved TX and RX then it's somewhat useless. If it's needed we may want to add comments on why the code is written this way.

These two patches work fine on Up2 but additional tests would be necessary to check if indeed they are functionally correct on additional APL/GLK platforms

### CODE REVIEW

**Comment by keyonjie** [comment]
I think re-enable HDA_DSP_ADSPIC_IPC in cmd_done() is not good, it will prevent all other IPC interrupts when we are in the middle of handling(including thread handling) previous IPC interrupt, this may decrease the efficiency especially at massive FW initiated messages sending case and lead to IPC timeout.

Please run stress test on IPC(e.g. 3 pipelines simultaneous with scripts in sof/tools/test/qa/tests/stability_test/multiplex_pipeline-test.sh), you will see that.

**Comment by RanderWang** [comment]
@plbossart  I port your PR to CNL, and it works.

Host would send a few IPC messages to DSP at setup time of a audio application.
DSP would send IPC message to host to update stream position pointer or DMA trace position.
So, if there are a few audio applications run in short time, this PR may add a little latency for
pointer or DMA trace updating.  I think the latency is tiny and could be ignored.  
I test playback and capture simultaneously and find no issue.

As I know, Japanese may care about this latency. :)

**Comment by plbossart** [comment]
On 2/22/19 2:09 AM, Keyon wrote:
>
> I think re-enable HDA_DSP_ADSPIC_IPC in cmd_done() is not good, it 
> will prevent all other IPC interrupts when we are in the middle of 
> handling(including thread handling) previous IPC interrupt, this may 
> decrease the efficiency especially at massive FW initiated messages 
> sending case and lead to IPC timeout.
>
The efficiency is limited by the two doorbell bits exposed by the 
hardware. You can only send one message at a time and will have to wait 
until the relevant DONE/BUSY bit is cleared by the other side AND it's 
able to deal with a new interrupt.

So no, sorry, enabling the interrupt earlier cannot possible make things 
more efficient. There is no real difference between unmasking an 
interrupt and enabling this interrupt really - I challenge you to prove 
me wrong here.

The only thing it can do is help deal with RX and TX threads in 
parallel, but only if the firmware does handle both channels independently.

> Please run stress test on IPC(e.g. 3 pipelines simultaneous with 
> scripts in 
> sof/tools/test/qa/tests/stability_test/multiplex_pipeline-test.sh), 
> you will see that.
>
I will, but please answer to my technical challenge above :-)
>

**Comment by plbossart** [comment]
> Please run stress test on IPC(e.g. 3 pipelines simultaneous with 
> scripts in 
> sof/tools/test/qa/tests/stability_test/multiplex_pipeline-test.sh), 
> you will see that.
>
Broken script, see https://github.com/thesofproject/sof/issues/1039

**Comment by plbossart** [comment]
@keyonjie see https://github.com/thesofproject/linux/issues/669, even with the existing IPC we can't do more than 3 streams, so any claims about efficiency are either incorrect or masked by other issues anyways...

**Comment by keyonjie** [comment]
> On 2/22/19 2:09 AM, Keyon wrote: I think re-enable HDA_DSP_ADSPIC_IPC in cmd_done() is not good, it will prevent all other IPC interrupts when we are in the middle of handling(including thread handling) previous IPC interrupt, this may decrease the efficiency especially at massive FW initiated messages sending case and lead to IPC timeout.
> The efficiency is limited by the two doorbell bits exposed by the hardware. You can only send one message at a time and will have to wait until the relevant DONE/BUSY bit is cleared by the other side AND it's able to deal with a new interrupt. So no, sorry, enabling the interrupt earlier cannot possible make things more efficient. There is no real difference between unmasking an interrupt and enabling this interrupt really - I challenge you to prove me wrong here. The only thing it can do is help deal with RX and TX threads in parallel, but only if the firmware does handle both channels independently.

What I meant by "decrease the efficiency" is exactly about RX and TX treads in parallel, imagine enabling travev for DMA trace inside FW, we can have 2 kinds of efficiency decreasing here:

a: Delivery of TX done ack is delayed so next host initiated IPC is also delayed.
1. host->DSP IPC A sent, waiting for done.
2. another thread want to send host->DSP IPC B, be blocked and waiting.
3. an FW initiated DMA trace arrived, schedule a rx_kwork to handle it.
4.  FW handled IPC A in 1, and try to send done to host, but IPC interrupt is masked.
5. the rx_kwork is scheduled, DMA trace IPC message is handled.
6. done interrupt for IPC A can be sent to host now.
7. host handle done interrupt handled, then IPC B can be sent.
8. considering done ack for IPC B can be delayed again, so here IPC B can have higher possibility to be timed out.

b: Delivery of FW initiated IPC might be dropped.
1. host->DSP IPC A sent, waiting for done.
2.  FW handled IPC A in 1, and send done to host.
3. 3a.  host handling this done ack to IPC A.
3b.  during 3a, A FW initiated DMA trace want to be sent, but IPC interrupt is disabled at 3a, waiting.
4.  3a done, then 3b be sent to host, host schedule a rx_kwork to handle 3b IPC.
5.  another FW initiated DMA trace want to be sent, but as 3b is not handled yet, no destination free buffer, dropped.

At this case, if we re-enable IPC interrupt in 3a earlier, we can have lower possibility of dropping IPC in 5.


> Please run stress test on IPC(e.g. 3 pipelines simultaneous with scripts in sof/tools/test/qa/tests/stability_test/multiplex_pipeline-test.sh), you will see that.
> I will, but please answer to my technical challenge above :-)
> […](#)

**Comment by keyonjie** [comment]
> @keyonjie see #669, even with the existing IPC we can't do more than 3 streams, so any claims about efficiency are either incorrect or masked by other issues anyways...

Haven't tried the latest sof-dev for weeks, @keqiaozhang and me tried it months back and 4(at least 3) streams worked at that moment.

Please try enabling tracev in FW if you can run 3 streams.

**Comment by ranj063** [comment]
@plbossart the first commit looks good. you are reverting this commit from keyon 
b50619398 . These checks are also in the FW and I think they are redundant there too. 


As for the second commit, I think there is a possibility of some FW initiated IPC's to get dropped/delayed which currently probably only affects trace. But given that we set the ipc timeout to 300ms and the IPC processing deadline to 100ms in the fw, it might be very difficult to hit a real issue.

**Comment by plbossart** [comment]
> @plbossart the first commit looks good. you are reverting this commit from keyon
> [b506193](https://github.com/thesofproject/linux/commit/b50619398c7befa3bc19266c352d0bf4cc51d948) . These checks are also in the FW and I think they are redundant there too.

Thanks Ranjani for the pointer. @keyonjie since you added this code can you please comment on why it's really needed. I don't understand your explanations at all.

> As for the second commit, I think there is a possibility of some FW initiated IPC's to get dropped/delayed which currently probably only affects trace. But given that we set the ipc timeout to 300ms and the IPC processing deadline to 100ms in the fw, it might be very difficult to hit a real issue.

yes, so far I only hit issues that are already hit with the existing solution, so it's anyone's guess what this 'smarter' way of dealing with interrupts actually provides. I am really not happy about the team dropping the ball on stress tests on the core IPC parts, now this effectively gates progress since we are shooting in the dark.

**Comment by plbossart** [comment]
closing since @lyakh is working on an improved IPC
